language/stdlib/modules/libra_system.move

```diff
@@ -0,0 +1,650 @@
+        };
+        if (&fullnodes_network_identity_pubkey != &discovery_info.fullnodes_network_identity_pubkey) {
+            *&mut discovery_info.fullnodes_network_identity_pubkey = fullnodes_network_identity_pubkey;
+            changed = true;
+        };
+        if (&fullnodes_network_address != &discovery_info.fullnodes_network_address) {
+            *&mut discovery_info.fullnodes_network_address = fullnodes_network_address;
+            changed = true;
+        };
+
+        changed
+    }
+
+    // Create a DiscoveryInfo from the ValidatorConfig stored at addr.
+    // Aborts if addr does not have a ValidatorConfig
+    fun make_discovery_info(addr: address): DiscoveryInfo {
+        let config = ValidatorConfig::config(addr);
+
+       DiscoveryInfo {
+           addr: addr,
+           validator_network_identity_pubkey:
+               ValidatorConfig::validator_network_identity_pubkey(&config),
+           validator_network_address:
+               ValidatorConfig::validator_network_address(&config),
+           fullnodes_network_identity_pubkey:
+               ValidatorConfig::fullnodes_network_identity_pubkey(&config),
+           fullnodes_network_address:
+               ValidatorConfig::fullnodes_network_address(&config),
+       }
+    }
+
+    // Trigger a reconfiguation the Libra system by:
+    // (1) Checking if there have been any additions since the last reconfiguration (and adding
+    //     the validator's config if so)
+    // (2) Checking if there have been any removals since the last reconfiguration (and removing
+    //     the validator if so)
+    // (3) Checking if there have been any key rotations since the last reconfiguration (and
+    //     updating the key info if so)
+    // (4) Emitting an event containing new validator set or discovery set, which will be
+    //     passed to the executor
+    fun reconfigure() acquires ValidatorSet, DiscoverySet {
+        // For now, this only supports a simple form of reconfiguration: allowing a fixed set of
+        // validators to rotate their keys.
+        // TODO: support adding and removing validators. Eventually, we will do this by computing
+        // the new validator set from a larger list of candidate validators sorted by stake.
+        let validator_set_ref = borrow_global_mut<ValidatorSet>(0x1D8);
+        let additions_ref = &mut validator_set_ref.additions;
+        let removals_ref = &mut validator_set_ref.removals;
+        let validators_vec_ref = &mut validator_set_ref.validators;
+        let discovery_set_ref = borrow_global_mut<DiscoverySet>(0xD15C0);
+        let discovery_vec_ref = &mut discovery_set_ref.discovery_set;
+        let validator_set_changed = false;
+        let discovery_set_changed = false;
+
+        // (1) Check for additions
+        let i = 0;
+        let len = Vector::length(additions_ref);
+        // if additions is nonempty, we have an addition to process
+        if (len > 0) {
+            loop {
```

Comment:
I understand we want to keep things as close as to what they were for now but I wonder if we should consider rewriting all those loops with a `while` instead of an unconditional `loop` at some point.
It could make the code a bit more elegant and readable...
thoughts?

Reply:
Agreed! The plan on this PR is to make the conversion stupidly straightforward. That way if there are any issues later on it should be easier to verify that the error didn't stem from a mis-port. There _will_ be a subsequent PR that Move-ifies these modules and fixes stuff like this.